single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,751 @@
+        )
+
+        container = (
+            self.charm.unit.get_container(CONTAINER) if self.substrate == Substrates.K8S else None
+        )
+
+        # Defined workloads and configs
+        self.define_workloads_and_config_managers(container)
+
+        # Managers
+        self.backup_manager = BackupManager(
+            self.charm,
+            self.role,
+            self.substrate,
+            self.state,
+            container,
+        )
+        self.tls_manager = TLSManager(
+            self,
+            self.workload,
+            self.state,
+            self.substrate,
+        )
+        self.mongo_manager = MongoManager(
+            self,
+            self.workload,
+            self.state,
+            self.substrate,
+        )
+
+        # Event Handlers
+        self.password_actions = PasswordActionEvents(self)
+        self.backup_events = BackupEventsHandler(self)
+        self.tls_events = TLSEventsHandler(self)
+        self.primary_events = PrimaryActionHandler(self)
+        self.client_events = DatabaseEventsHandler(self, RelationNames.DATABASE)
+
+    @property
+    def config(self):
+        """Returns the actual config."""
+        return self.charm.parsed_config
+
+    def define_workloads_and_config_managers(self, container: Container | None) -> None:
+        """Export all workload and config definition for readability."""
+        # BEGIN: Define workloads.
+        self.workload = get_mongodb_workload_for_substrate(self.substrate)(
+            role=self.role, container=container
+        )
+        self.mongos_workload = get_mongos_workload_for_substrate(self.substrate)(
+            role=self.role, container=container
+        )
+        # END: Define workloads
+
+        # BEGIN Define config managers
+        self.config_manager = MongoDBConfigManager(
+            self.config,
+            self.state,
+            self.workload,
+        )
+        self.mongos_config_manager = MongosConfigManager(
+            self.config,
+            self.mongos_workload,
+            self.state,
+        )
+        self.logrotate_config_manager = LogRotateConfigManager(
+            self.role,
+            self.substrate,
+            self.config,
+            self.state,
+            container,
+        )
+        self.mongodb_exporter_config_manager = MongoDBExporterConfigManager(
+            self.role,
+            self.substrate,
+            self.config,
+            self.state,
+            container,
+        )
+        # END: Define config managers
+
+    @property
+    def config_managers(self) -> Iterable[CommonConfigManager]:  # pragma: nocover
+        """All config managers for iteration."""
+        return (
+            self.config_manager,
+            self.mongos_config_manager,
+            self.backup_manager,
+            self.logrotate_config_manager,
+            self.mongodb_exporter_config_manager,
+        )
+
+    # BEGIN: Handlers.
+
+    @override
+    def on_install(self) -> None:
```

Comment:
It's specifically to keep with the structure that they are here because the other operator will have different handlers for those methods.
There is a single `LifecycleEventHandlers` which is able to work with the two operators (MongoDB and Mongos) without having to know which operator is provided.
So it's the responsibility of the Operator to act accordingly. This keeps a single reponsibility principle.

Reply:
This is helpful! Thank you a lot! I think my concerns about:

> might we want all of these event handler functions in the directory "event"

will be quelled from when this file is moved into a directory that is separate from the mangers like was mentioned earlier